ilqr.py

```diff
@@ -343,17 +366,328 @@ def _forward_pass(self, L_last):
             ls_iters:   Number of linesearch iterations
         """
         # Do linesearch to determine eps
+        timeStart = time.time()
         eps, x, u, L, ls_iters = self._linesearch(L_last)
+        timeEnd = time.time()
+        self.time_fp = timeEnd - timeStart
 
-        # Compute the first-order dynamics derivatives
-        for t in range(0,self.N-1):
-            self.fx[:,:,t], self.fu[:,:,t] = self._calc_dynamics_partials(x[:,t], u[:,t])
+        timeStart = time.time()
+        self._get_derivatives(x, u)
+        timeEnd = time.time()
+        self.time_getDerivs = timeEnd - timeStart
 
         # Update stored values
         self.u_bar = u
         self.x_bar = x
 
         return L, eps, ls_iters
+
+    def _get_derivatives(self, x, u):
+        """
+        Calculates the derivatives fx and fu over the entire trajectory. Depending on 
+        the keypoint method specified, this function will calculate a set of keypoints.
+        At these keypoints the derivatives will be calculated exactly using autodiff.
+        In-between these keypoints this function will linearly interpolate approximations
+        to the derivatives between the computed values. This is done to reduce the amount
+        of time spent per iteration computing derivatives.
+
+        Updates:
+            fx:      dynamcis partial wrt state at each timestep
+            fu:      dynamcis partial wrt control at each timestep
+        """
+
+        # Debug variable to plot some useful graphs if required, also calculates the 
+        # derivatives at every time step and compute the error of the interpolation
+        # for debugging
+        DEBUG = False
```

Comment:
I would do something similar if I were developing new code, but let's get rid of all of these `DEBUG` blocks before merging into the main branch. I'm sure the plots and errors are good info, but I'm just not sure that any one other than you would know what they mean.